// -----------------------------------------------------------------------------
// Copyright 2018 Frank V. Castellucci and Arthur Greef
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// -----------------------------------------------------------------------------

syntax = "proto3";
option java_multiple_files = true;
option java_package = "hashblock.asset.protobuf";

// On chain resource

message Property {
    string name = 1;
    string value = 2;
}

message Asset {
    string system = 1;
    string key = 2;
    string value = 3;
    repeated Property properties = 4;
}

// Asset Payload
// - Contains a proposal, vote or asset (for direct and genesis).

message AssetPayload {
    // The action indicates data is contained within this payload
    enum Action {
        ACTION_UNSET = 0;

        // A proposal action - data will be a AssetProposal
        ACTION_PROPOSE = 1;

        // A vote action - data will be a AssetVote
        ACTION_VOTE = 2;

        // New 0.2.0: Genesis action - data will be Asset
        ACTION_GENESIS = 3;

        // New 0.3.0: Genesis action - data will be Asset
        ACTION_DIRECT = 4;
    }
    // The action of this payload
    Action action = 1;

    // The content of this payload
    bytes data = 2;
}

// Asset Proposal
//
// This message proposes an Asset.

message AssetProposal {
    // The asset being proposed
    bytes asset = 1;

    // allow duplicate proposals with different hashes
    // randomly created by the client

    string nonce = 2;
}

// Asset Vote
//
// This message votes on Asset Proposal

message AssetVote {
    enum Vote {
        VOTE_UNSET = 0;
        VOTE_ACCEPT = 1;
        VOTE_REJECT = 2;
    }

    // The id of the proposal being voted on

    string proposal_id = 1;

    Vote vote = 2;
}


// Contains the vote counts for a given proposal.

message AssetCandidate {
    // An individual vote record
    message VoteRecord {
        enum Vote {
            VOTE_UNSET = 0;
            VOTE_ACCEPT = 1;
            VOTE_REJECT = 2;
        }
        // The public key of the voter
        string public_key = 1;

        // The voter's actual vote
        Vote vote = 2;

    }

    // The proposal id, a hash of the original proposal
    string proposal_id = 1;

    // The active proposal itself
    AssetProposal proposal = 2;

    // list of votes
    repeated VoteRecord votes = 3;
}

// Contains all the setting candiates up for vote.

message AssetCandidates {
    repeated AssetCandidate candidates = 1;
}

